packages/styled-components/src/sheet/Rehydration.js

```diff
@@ -5,7 +5,7 @@ import { getIdForGroup, setGroupForId } from './GroupIDAllocator';
 import type { Sheet } from './types';
 
 const SELECTOR = `style[${SC_ATTR}][${SC_ATTR_VERSION}="${SC_VERSION}"]`;
-const RULE_RE = /(?:\s*)?(.*?){((?:{[^}]*}|(?!{).*?)*)}/g;
+const RULE_RE = /([^{]*){((?:[^{}]+(?:{[^}]*})?)*)}/g;
```

Comment:
- Leading `(?:\s*)?` is gone. This isn't needed because `exec` will keep walking forwards character by character. Also we want to start matching as soon as possible and if we later don't match, then this will slow the regex down because it'll start backtracking.
- Replaced `.*?` (btw `*?` is redundant) with `[^{]*` because we know that a `{` must follow. So we safe some time here by telling the Regex that we're matching as much as possible until that char instead of eagerly matching when the following characters don't quite match.
- Internally in the regex we had the same problems but more pronounced.

Reply:
This doesn't work for certain types of deeply nested media queries unfortunately, use this as a test string:

```
@media (max-width:300px){.fvsNoc{display:block;}@supports (display:-webkit-box) or (display:-webkit-flex) or (display:-ms-flexbox) or (display:flex){.fvsNoc{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}}@media (max-width:320px){.fvsNoc{display:block;}}}
data-styled.g1[id="pages__Container-sc-4wa63o-0"]{content:"fvsNoc,"}
.dBiKfw{font-size:8rem;font-weight:900;color:darkgreen;text-align:center;}
data-styled.g2[id="pages__Heading-sc-4wa63o-1"]{content:"dBiKfw,"}
```

(it's from the original issue I created the regex from)

The revised regex breaks at the second `@media`

Reply:
My suggestion here is to change the SSR emitter to add a special character CSS doesn't process to serve as the "split" point (for use in `string.split()` and just ditch the splitting regex.

Reply:
We could likely just add `/*!sc*/` as splitting markers